pkg/ipamd/datastore/data_store.go

```diff
@@ -985,10 +986,14 @@ func (ds *DataStore) GetENINeedsIP(maxIPperENI int, skipPrimary bool) *ENI {
 		if len(eni.AvailableIPv4Cidrs) < maxIPperENI {
 			ds.log.Debugf("Found ENI %s that has less than the maximum number of IP/Prefixes addresses allocated: cur=%d, max=%d",
 				eni.ID, len(eni.AvailableIPv4Cidrs), maxIPperENI)
-			return eni
+			enis = append(enis, eni)
+			if useSubnetDiscovery {
```

Comment:
Wondering should we even do this? we could always get all ENIs and for non-subnet-Discovery scenario we just pick eni[0] in the caller.. also running the loop should be fine since the number of ENis is always capped by the max enis supported per instance..

Reply:
Either seems fine with me. The `useSubnetDiscovery` seems to scope operation of getting all subnets to the specific feature, which in other words, closer to previous behavior.

Reply:
Now that I think about it we don't need this conditional.. Non-subnet discovery would only have one unfilled ENI at a time since we only have one subnet. We can just return all unfilled ENIs

Reply:
Either seems fine. 

> running the loop should be fine since the number of ENis is always capped by the max enis supported per instance

And we aren't doing any API call here. So, seems fine.

Reply:
In the case of custom networking removing this conditional would work as well (e.g. primary ENI is unfilled and subnet is exhausted, secondary ENI has space for more IPs and is allocated on subnet from ENIConfig). The caller would attempt to allocate IPs on primary ENI --> fail --> attempt on secondary ENI (ENIConfig Subnet)